### PR TITLE
SCRD-2813 Add support for CPI parameters

### PR DESCRIPTION
**Pull request by aagate**
* Create /etc/caasp/cpi/openstack.conf on the admin
  node with CPI parameters
* Set the CPI parameter subnet-id from caasp subnet heat resource
* Set other CPI parameters like: auth-url, tenant-name,
  region, username, password, monitor-max-retries, bs-version,
  and ignore-vol-az from heat parameters in caasp-environment.yaml
* Modified both caasp-stack.yaml and caasp-multi-master-stack.yaml
  with this change.

### CODE REVIEW

File: caasp-environment.yaml
Comment (mjura) [inline caasp-environment.yaml]: This flavor is too small for CaaSP nodes :/

File: caasp-environment.yaml
Comment (mjura) [inline caasp-environment.yaml]: In `caasp-stack.yaml` tempate it is `cpi_cinder_bs_version`

File: caasp-stack.yaml
Comment (mjura) [inline caasp-stack.yaml]: Valid values are v1, v2, v3 and auto

File: caasp-environment.yaml
Comment (mjura) [reply caasp-environment.yaml]: Valid values are v1, v2, v3 and auto

**Review by mjura — CHANGES_REQUESTED** [review]
Changes are needed

File: caasp-environment.yaml
Comment (mjura) [inline caasp-environment.yaml]: In caasp-stack.yaml it is defined as `cpi_ignore_vol_az:`

File: caasp-stack.yaml
Comment (mjura) [inline caasp-stack.yaml]: Please remove spaces `bs-version=$cpi_cinder_bs_version`

File: caasp-stack.yaml
Comment (mjura) [inline caasp-stack.yaml]: Please remove spaces `ignore-vol-az=$cpi_ignore_vol_az`

**Review by mjura — CHANGES_REQUESTED** [review]
Some changes are needed

**Comment by mjura** [comment]
Please rebase

File: caasp-stack.yaml
Comment (mjura) [inline caasp-stack.yaml]: Here should be `floating-network-id=$cpi_floating_id`

File: caasp-stack.yaml
Comment (mjura) [reply caasp-stack.yaml]: It should be `ignore-volume-az=$cpi_ignore_vol_az`

**Review by mjura — CHANGES_REQUESTED** [review]
Changes needed

**Comment by mjura** [comment]
This requires https://github.com/kubic-project/velum/pull/676

File: caasp-environment.yaml
Comment (aagate) [reply caasp-environment.yaml]: done

File: caasp-environment.yaml
Comment (aagate) [reply caasp-environment.yaml]: thx! done added constraint, fixed parameter name

File: caasp-stack.yaml
Comment (aagate) [reply caasp-stack.yaml]: thx! done added constraint, fixed parameter name

File: caasp-stack.yaml
Comment (aagate) [reply caasp-stack.yaml]: done

File: caasp-stack.yaml
Comment (aagate) [reply caasp-stack.yaml]: outdated

File: caasp-stack.yaml
Comment (aagate) [reply caasp-stack.yaml]: thx! done

**Comment by aagate** [comment]
Rebased. Need https://github.com/kubic-project/velum/pull/676 to be available in CaaSP3 GM image before this can be merged, otherwise admin node UI does not come up.

File: caasp-stack.yaml
Comment (mjura) [inline caasp-stack.yaml:124]: `type: boolean`

File: caasp-stack.yaml
Comment (mjura) [inline caasp-stack.yaml]: default: true

**Review by mjura — COMMENTED** [review]
Only some nitpicks

File: caasp-stack.yaml
Comment (mjura) [inline caasp-stack.yaml]: Here should be `monitor-max-retries=$cpi_lb_mon_retries`

File: caasp-environment.yaml
Comment (mjura) [inline caasp-environment.yaml]: Example could be `https://api.keystone.example.net:5000/v3`

File: caasp-environment.yaml
Comment (mjura) [inline caasp-environment.yaml]: Please rename it to `cpi_bs_version`

File: caasp-stack.yaml
Comment (mjura) [inline caasp-stack.yaml]: Please rename it to `cpi_bs_version`

File: caasp-stack.yaml
Comment (mjura) [inline caasp-stack.yaml]: Please rename it to `cpi_bs_version`

File: caasp-stack.yaml
Comment (mjura) [inline caasp-stack.yaml]: Cloud Provider Interface - Keystone API URL

File: caasp-stack.yaml
Comment (mjura) [inline caasp-stack.yaml]: OpenStack Cloud Provider Integration - Keystone domain name

File: caasp-stack.yaml
Comment (mjura) [inline caasp-stack.yaml]: OpenStack Cloud Provider Integration - Keystone project name

File: caasp-stack.yaml
Comment (mjura) [inline caasp-stack.yaml]: OpenStack Cloud Provider Integration - Keystone region

File: caasp-stack.yaml
Comment (mjura) [inline caasp-stack.yaml]: OpenStack Cloud Provider Integration - Keystone username

File: caasp-stack.yaml
Comment (mjura) [inline caasp-stack.yaml]: OpenStack Cloud Provider Integration - Keystone user password

File: caasp-stack.yaml
Comment (mjura) [inline caasp-stack.yaml]: OpenStack Cloud Provider Integration - Neutron Load Balancer monitoring max retries

File: caasp-stack.yaml
Comment (mjura) [inline caasp-stack.yaml]: OpenStack Cloud Provider Integration - Cinder block storage API version

File: caasp-stack.yaml
Comment (mjura) [inline caasp-stack.yaml]: OpenStack Cloud Provider Integration - ignore Nova and Cinder availability zone
￼

File: caasp-stack.yaml
Comment (mjura) [inline caasp-stack.yaml]: Sorry, please revert it to `type: string`

File: caasp-stack.yaml
Comment (mjura) [inline caasp-stack.yaml]: Please change it to `bs-version=$cpi_bs_version`

File: caasp-stack.yaml
Comment (mjura) [inline caasp-stack.yaml]: Please change it to `$cpi_monitor_max_retries: { get_param: cpi_monitor_max_retries }`

File: caasp-stack.yaml
Comment (mjura) [inline caasp-stack.yaml]: `$cpi_bs_version: { get_param: cpi_bs_version }`

**Review by mjura — CHANGES_REQUESTED** [review]
Still some changes needed

**Comment by mjura** [comment]
I've submitted another fix to Velum
https://github.com/kubic-project/velum/pull/683

**Comment by aagate** [comment]
Thx for the review comments @mjura. Do you know when will the velum fixes make it to CaaSP Openstack Image? Or in the mean time can I do something on admin node to apply velum fixes manually? Right now it feels like I am flying blind since I cannot tryout and test these changes since admin node ui doesnt comeup

**Comment by mjura** [comment]
@aagate: Fixes are currently in CaaSP devel image, please ping me on IRC and I will past you in links to it.

**Comment by aagate** [comment]
Thanks @mjura! Posted a message in  #caasp channel on IRC

File: caasp-environment.yaml
Comment (mjura) [inline caasp-environment.yaml]: Default value should be `RegionOne` or `CustomRegion`

File: caasp-multi-master-environment.yaml
Comment (mjura) [inline caasp-multi-master-environment.yaml]: Default value should be `RegionOne` or `CustomRegion`

File: caasp-multi-master-stack.yaml
Comment (mjura) [inline caasp-multi-master-stack.yaml]: Default value should be `RegionOne` or `CustomRegion`

File: caasp-stack.yaml
Comment (mjura) [inline caasp-stack.yaml]: Default value should be `RegionOne` or `CustomRegion`

File: caasp-stack.yaml
Comment (mjura) [inline caasp-stack.yaml]: Please put this string into quotation marks `"true"`

File: caasp-multi-master-stack.yaml
Comment (mjura) [inline caasp-multi-master-stack.yaml]: Please put this string into quotation marks `"true"`

File: caasp-multi-master-environment.yaml
Comment (mjura) [inline caasp-multi-master-environment.yaml]: Please put this string into quotation marks `"true"`

File: caasp-environment.yaml
Comment (mjura) [inline caasp-environment.yaml]: Please put this string into quotation marks `"true"`

**Review by mjura — CHANGES_REQUESTED** [review]
Still changes are needed

File: caasp-environment.yaml
Comment (aagate) [reply caasp-environment.yaml]: done

File: caasp-multi-master-environment.yaml
Comment (aagate) [reply caasp-multi-master-environment.yaml]: done

File: caasp-multi-master-stack.yaml
Comment (aagate) [reply caasp-multi-master-stack.yaml]: done

File: caasp-stack.yaml
Comment (aagate) [reply caasp-stack.yaml]: done

File: caasp-stack.yaml
Comment (aagate) [reply caasp-stack.yaml]: done

File: caasp-multi-master-stack.yaml
Comment (aagate) [reply caasp-multi-master-stack.yaml]: done

File: caasp-multi-master-environment.yaml
Comment (aagate) [reply caasp-multi-master-environment.yaml]: done

File: caasp-environment.yaml
Comment (aagate) [reply caasp-environment.yaml]: done

**Review by mjura — DISMISSED** [review]
LGTM

**Comment by aagate** [comment]
I think this is good to be merged. I have tested with CaaSP 3.0 image and Maintenance Update (MU) 9687 https://maintenance.suse.de/incident/9687. 
Also it is now necessary to create a cluster with CaaSP 3.0 image AND  the MU. Without the fixes to the velum dashboard that were available in the MU the admin node wont come up.

**Comment by aagate** [comment]
@mjura @sayalilunkad @rsalevsky  can you please review this and approve if possible? thank you!

File: caasp-multi-master-stack.yaml
Comment (dirkmueller) [inline caasp-multi-master-stack.yaml:123]: Why are we setting the default to something that is always going to be wrong?

File: caasp-multi-master-stack.yaml
Comment (dirkmueller) [inline caasp-multi-master-stack.yaml]: Why? Can this be made slightly less insecure?

**Review by dirkmueller — DISMISSED** [review]
At least the 0777 is probably not a good thing to merge. There were also comments from @mjura that didn't get addressed

File: caasp-environment.yaml
Comment (aagate) [reply caasp-environment.yaml]: done. outdated

File: caasp-stack.yaml
Comment (aagate) [reply caasp-stack.yaml]: done. outdated

File: caasp-stack.yaml
Comment (aagate) [reply caasp-stack.yaml:124]: outdated

File: caasp-stack.yaml
Comment (aagate) [reply caasp-stack.yaml]: outdated

File: caasp-environment.yaml
Comment (aagate) [reply caasp-environment.yaml]: done. outdated

File: caasp-stack.yaml
Comment (aagate) [reply caasp-stack.yaml]: done. outdated

File: caasp-environment.yaml
Comment (aagate) [reply caasp-environment.yaml]: done. outdated

File: caasp-stack.yaml
Comment (aagate) [reply caasp-stack.yaml]: done. outdated

File: caasp-stack.yaml
Comment (aagate) [reply caasp-stack.yaml]: done outdated

File: caasp-stack.yaml
Comment (aagate) [reply caasp-stack.yaml]: done outdated

File: caasp-stack.yaml
Comment (aagate) [reply caasp-stack.yaml]: done outdated

File: caasp-stack.yaml
Comment (aagate) [reply caasp-stack.yaml]: done outdated

File: caasp-multi-master-stack.yaml
Comment (aagate) [reply caasp-multi-master-stack.yaml:123]: The user still has chance to change it to the appropriate one, during the  CaaSP admin  node bootstrapping step (its available as a text field in the UI, there is a hint as to where the user can go and get the appropriate endpoint, and is mentioned in the docs as well). Unfortunately heat does not currently give access to this pubilc endpoint, unless the endpoint is created by the heat template, so we cannot set it automagically.

File: caasp-multi-master-stack.yaml
Comment (aagate) [reply caasp-multi-master-stack.yaml]: Looking into these permissions. 

All the comments by @mjura were addressed already,  marked them as "done outdated"

**Comment by aagate** [comment]
Still working to see if we can tighten permissions on /etc/caasp/cpi/openstack.conf

We are are now  blocked by this CaaSP bug https://bugzilla.suse.com/show_bug.cgi?id=1122439, will have to wait for bugfix to make its way to a MU before we can make progress on this review.

File: caasp-multi-master-stack.yaml
Comment (aagate) [reply caasp-multi-master-stack.yaml]: now set permissions to 0755 on the /etc/caasp/cpi directory (same as other dirs in that caasp folder) and 0644 on /etc/caasp/cpi/openstack.conf

File: caasp-stack.yaml
Comment (aagate) [inline caasp-stack.yaml:290]: set permissions to 0755 on /etc/caasp/cpi directory

File: caasp-stack.yaml
Comment (aagate) [inline caasp-stack.yaml:294]: set permissions to 0644 on /etc/caasp/cpi/openstack.conf directory

File: caasp-multi-master-stack.yaml
Comment (aagate) [inline caasp-multi-master-stack.yaml:324]: set permissions to 0755 on /etc/caasp/cpi directory

File: caasp-multi-master-stack.yaml
Comment (aagate) [inline caasp-multi-master-stack.yaml:328]: set permissions to 0644 on /etc/caasp/cpi/openstack.conf

**Review by aagate — COMMENTED** [review]
set permissions to 0755 on /etc/caasp/cpi directory and 0644 on /etc/caasp/cpi/openstack.conf file

**Comment by spiarh** [comment]
@aagate what is the idea behind writing the parameters in `/etc/caasp/cpi/openstack.conf` ?

**Comment by aagate** [comment]
@lcavajani  By writing  to /etc/caasp/cpi/openstack.conf certain openstack parameters would be pre-populated in the Cloud Provider Integration section of the velum dashboard during the bootstrap stage

**Comment by spiarh** [comment]
@aagate ok thanks, I was not aware of that feature

**Comment by aagate** [comment]
@mjura @lcavajani @rsalevsky can you please review and approve this PR if possible ? thank you!

**Review by rsalevsky — APPROVED** [review]

**Pull request merged**
